File: src/y2021/02. Dive!.py
#!/usr/bin/env python3

"""
AoC Day 2 - Dive! - in Python.
"""

__author__ = "Lucas Moeskops"
__date__ = "2021-12-02"

from sys import stdin


LINES = stdin.read().split('\n')

commands = [(line.split(' ')[0], int(line.split(' ')[1])) for line in LINES]


def part_1():
    position, depth = 0, 0
    for direction, amount in commands:
        match direction:
            case 'forward':
                position += amount
            case 'up':
                depth -= amount
            case 'down':
                depth += amount
    return position * depth


def part_2():
    aim, position, depth = 0, 0, 0
    for direction, amount in commands:
        match direction:
            case 'forward':
                position += amount
                depth += amount * aim
            case 'up':
                aim -= amount
            case 'down':
                aim += amount
    return position * depth


print(part_1())
print(part_2())

File: src/y2021/02. Dive!.hs
{- AoC Day 2 - Dive! - in Haskell.
 -
 - Author: Lucas Moeskops
 - Date: 2021-12-02
 -}


task_1 = run 0 0
    where
        run position depth []                      = position * depth
        run position depth ((direction,amount):cs) =
            case direction of
                "forward" -> run (position + amount) depth            cs
                "up"      -> run position            (depth - amount) cs
                "down"    -> run position            (depth + amount) cs
task_2 = run 0 0 0
    where
        run position depth aim []                      = position * depth
        run position depth aim ((direction,amount):cs) =
            case direction of
                "forward" -> run (position + amount) (depth + amount * aim) aim            cs
                "up"      -> run position            depth                  (aim - amount) cs
                "down"    -> run position            depth                  (aim + amount) cs

parse :: String -> [(String, Int)]
parse = map parseLine.lines
    where parseLine = (\(direction:amount:ws) -> (direction, read amount)).words

main = do
    commands <- parse <$> getContents
    putStrLn$(++) "[Part 1]: ".show$task_1 commands
    putStrLn$(++) "[Part 2]: ".show$task_2 commands
